Arrange configuration and motor tab

haslinghuis opened the pull request:
Fixes: #2304 

**Summary**: 
- [x] Move mixer and motor/esc features from configuration to motor tab
- [x] Move servo test to servos tab

**Suggestions**:
- [x] Styling changes notice, mixer and buttons @Asizon / @limonspb 
- [x] Fix `MOTOR_STOP` alignment with `ESC_SENSOR` and align padding on different places @TheIsotopes
- [x] Move reorder Motors and Motor direction to Mixer settings @zzyzx / @limonspb
- [x] Move mixer from graph to configuration and put configuration before testing @McGiverGim  
- [x] Stop motor testing when making configuration changes (+ disable motor test after making changes before reboot) @mikeller  

Thanks  @Asizon  @bizmar @chmelevskij @limonspb @McGiverGim @mikeller @TheIsotopes @zzyzx

chmelevskij commented on src/js/tabs/servos.js:
```diff
@@ -1,36 +1,30 @@
     }
-    get_servo_configurations();
+
+    Promise
+    .resolve(true)
+    .then(() => MSP.promise(MSPCodes.MSP_SERVO_CONFIGURATIONS))
+    .then(() => MSP.promise(MSPCodes.MSP_SERVO_MIX_RULES))
+    .then(() => MSP.promise(MSPCodes.MSP_RC))
+    .then(() => MSP.promise(MSPCodes.MSP_BOXNAMES))
+    .then(load_html)
+    .catch((error) => console.log(error.message));
```
Do these need to be run in parallel or they can be in sequence?

If we make `initialise` async we could do something like for the sequential
```javascript
try {
  await MSP.promise(MSPCodes.MSP_SERVO_CONFIGURATIONS);
  await MSP.promise(MSPCodes.MSP_SERVO_MIX_RULES);
  await MSP.promise(MSPCodes.MSP_RC);
  await MSP.promise(MSPCodes.MSP_BOXNAMES);
  load_html();
} catch(error) {
  console.log(error)
}
```

Or for parallel
```javascript
try {
  await Promise.all([
    MSP.promise(MSPCodes.MSP_SERVO_CONFIGURATIONS),
    MSP.promise(MSPCodes.MSP_SERVO_MIX_RULES),
    MSP.promise(MSPCodes.MSP_RC),
    MSP.promise(MSPCodes.MSP_BOXNAMES),
  ])
  load_html();
} catch(error) {
  console.log(error)
}
```

haslinghuis replied:
There is no particular sequence as they don't depend on each other. So the first solution would work but I'm in doubt in what advantage the parallel solution has over the former.

haslinghuis replied:
Thanks @chmelevskij - it's working 😜 Updated now in `servos.js` will change in `configuration.js` and `motors.js` also

haslinghuis replied:
Updated `motors.js` also. The initialize part is okay, but somehow the save function does not like the `async` approach. This snippet won't work even with `async function process_html() {` on line 222
```
$('a.save').on('click', async () => {
            // gather data that doesn't have automatic change event bound
            FC.MOTOR_CONFIG.minthrottle = parseInt($('input[name="minthrottle"]').val());
            FC.MOTOR_CONFIG.maxthrottle = parseInt($('input[name="maxthrottle"]').val());
            FC.MOTOR_CONFIG.mincommand = parseInt($('input[name="mincommand"]').val());

            if (semver.gte(FC.CONFIG.apiVersion, API_VERSION_1_42)) {
                FC.MOTOR_CONFIG.motor_poles = parseInt($('input[name="motorPoles"]').val());
            }

            FC.MOTOR_3D_CONFIG.deadband3d_low = parseInt($('input[name="3ddeadbandlow"]').val());
            FC.MOTOR_3D_CONFIG.deadband3d_high = parseInt($('input[name="3ddeadbandhigh"]').val());
            FC.MOTOR_3D_CONFIG.neutral = parseInt($('input[name="3dneutral"]').val());

            FC.PID_ADVANCED_CONFIG.fast_pwm_protocol = parseInt(esc_protocol_e.val() - 1);
            FC.PID_ADVANCED_CONFIG.use_unsyncedPwm = $('input[id="unsyncedPWMSwitch"]').is(':checked') ? 1 : 0;
            FC.PID_ADVANCED_CONFIG.motor_pwm_rate = parseInt($('input[name="unsyncedpwmfreq"]').val());
            FC.PID_ADVANCED_CONFIG.digitalIdlePercent = parseFloat($('input[name="digitalIdlePercent"]').val());

            if (semver.gte(FC.CONFIG.apiVersion, "1.25.0") && semver.lt(FC.CONFIG.apiVersion, API_VERSION_1_41)) {
                FC.PID_ADVANCED_CONFIG.gyroUse32kHz = $('input[id="gyroUse32kHz"]').is(':checked') ? 1 : 0;
            }

            analytics.sendChangeEvents(analytics.EVENT_CATEGORIES.FLIGHT_CONTROLLER, self.analyticsChanges);
            self.analyticsChanges = {};

            try {
                await MSP.promise(MSPCodes.MSP_SET_ARMING_CONFIG, mspHelper.crunch(MSPCodes.MSP_SET_ARMING_CONFIG));
                await MSP.promise(MSPCodes.MSP_SET_MIXER_CONFIG, mspHelper.crunch(MSPCodes.MSP_SET_MIXER_CONFIG));
                await MSP.promise(MSPCodes.MSP_SET_MOTOR_CONFIG, mspHelper.crunch(MSPCodes.MSP_SET_MOTOR_CONFIG));
                await MSP.promise(MSPCodes.MSP_SET_MOTOR_3D_CONFIG, mspHelper.crunch(MSPCodes.MSP_SET_MOTOR_3D_CONFIG));
                await MSP.promise(MSPCodes.MSP_SET_ADVANCED_CONFIG, mspHelper.crunch(MSPCodes.MSP_SET_ADVANCED_CONFIG));
                await MSP.promise(MSPCodes.MSP_SET_FILTER_CONFIG, mspHelper.crunch(MSPCodes.MSP_SET_FILTER_CONFIG));
                await MSP.promise(MSPCodes.MSP_SET_FEATURE_CONFIG, mspHelper.crunch(MSPCodes.MSP_SET_FEATURE_CONFIG));
                await MSP.promise(MSPCodes.MSP_EEPROM_WRITE);
                GUI.log(i18n.getMessage('configurationEepromSaved'));

                GUI.tab_switch_cleanup(() => {
                    await MSP.promise(MSPCodes.MSP_SET_REBOOT);
                });

                analytics.sendChangeEvents(analytics.EVENT_CATEGORIES.FLIGHT_CONTROLLER, self.analyticsChanges);
                self.analyticsChanges = {};
                reinitialiseConnection();
            } catch (error) {
                console.log(error);
            }
        });
```
It produces the infamous
```
main.js:362 Uncaught TypeError: Cannot read property 'initialize' of undefined
    at main.js:362
    at Object.TABS.setup.cleanup (setup.js:351)
    at GuiControl.tab_switch_cleanup (gui.js:278)
    at HTMLAnchorElement.<anonymous> (main.js:266)
    at HTMLAnchorElement.dispatch (jquery.min.js:2)
    at HTMLAnchorElement.v.handle (jquery.min.js:2)
DevTools failed to load SourceMap: Could not load content for chrome-extension://ajdoecjhplhoahggfmgflfnlechcljoc/node_modules/jquery-textcomplete/dist/dist/jquery.textcomplete.min.map: System 
```

chmelevskij replied:
If id doesn't matter about the sequence I'd use `Promise.all`, bunch of `await`s will block and wait for the previous to finish so you might get a bit of a lag if execution time adds up to 1s or smth

McGiverGim replied:
We added some time ago one MSP command to execute various MSP commands at once. I don't remember the ID but can be useful for that.

McGiverGim replied:
I found it: https://github.com/betaflight/betaflight-configurator/pull/1605

mikeller replied:
MSP commands can definitely not be run in parallel - the firmware and the protocol lack the means to sequentialise / recombine responses.

mikeller replied:
As for `MSP_MULTIPLE_MSP`, I'd be reluctant to use this as a wrapper to group multiple commands in configurator - this will break and require another rewrite as soon as one of the combined frames (request or response) exceeds 255 bytes in length, which will happen sooner or later due to the MSP protocol's nature of having commands grow in lenght with every change.

haslinghuis commented:
@Asizon FYI: the part that doesn't work in `servos.js` is this snippet:
```
    <div class="servos_testing">
        <div class="title2" i18n="servosText"></div>
        <ul class="titles">
            <li i18n_title="servoNumber8">8</li>
            <li i18n_title="servoNumber7">7</li>
            <li i18n_title="servoNumber6">6</li>
            <li i18n_title="servoNumber5">5</li>
            <li i18n_title="servoNumber4">4</li>
            <li i18n_title="servoNumber3">3</li>
            <li i18n_title="servoNumber2">2</li>
            <li i18n_title="servoNumber1">1</li>
        </ul>
        <div class="bar-wrapper"></div>
    </div>
```

mikeller commented:
@haslinghuis: Seeing how much change is still going on in this pull request, can you please change this to 'draft' status so reviewers can hold off until it is ready for review?

haslinghuis commented:
![Screenshot from 2020-12-27 00-55-09](https://user-images.githubusercontent.com/8344830/103161220-c4808180-47de-11eb-814a-d45d54f3063d.png)

haslinghuis commented:
@mikeller Sorry should have switched the status. But it's battle tested and ready for review now

haslinghuis commented:
I am really struggling with the bars in servos.js
Can't tell if it's an css/html or code issue.

McGiverGim commented:
@haslinghuis did you see my comment about MSP_MULTIPLE_MSP above?

haslinghuis commented:
@McGiverGim yes but there is no example how to call it. And I was very very busy trying to fix the servos test. Motor tab and configuration tab are ready and @Asizon offered to look at servos because I am on dead trail with it.

McGiverGim commented:
> @McGiverGim yes but there is no example how to call it. And I was very very busy trying to fix the servos test. Motor tab and configuration tab are ready and @Asizon offered to look at servos because I am on dead trail with it.

In the PR text there is one example. Very easy to use. I think is better to use it than to use a lot of parallel async calls: the FC has not parallel processing capabilities.

mikeller commented:
Not sure what we _need_ to improve in how the protocol is handled, except for the semantics of how commands are chained in configurator.
The fact that we are using organically grown commands to get / set configuration, and the fact that this is done on every tab load are both design flaws of the current design, and going to great lengths to optimise this flawed approach isn't going to fix the flaws.
We should better look into moving to an approach where the entire configuration is dumped in an efficient way on connect, and then written back at the very end based on what has been changed.

haslinghuis commented:
@McGiverGim I tried to use `MULTIPLE_MSP` but it's failing. So I guess I bring it back to original state. I like the load-all / save-all approach (like BIOS).
Also decided to split the PR. Moving servo test bars requires much CSS rework and need help there so I left this part out for a follow up PR. Only have to finalize the MSP part and it should be alright.

chmelevskij commented:
I'm just wondering what's the advantage of `MULTIPLE_MSP` for this usecase? Since only thing is cared about here is that it happens and ui updated, no response needed, using plain `Promise.all` means anyone seeing the code will be able to know what roughly is happening, with custom handlers it adds some abstraction overhead, where you have to know how that will work.

haslinghuis commented:
@chmelevskij added my own alternative (`Promise.all` or `MULTIPLE_MSP` all have same outcome) - we should rather focus on an BIOS like approach (Load once, save once, only have to deal with status polls)

TheIsotopes commented:
@haslinghuis found two bugs on Motors tab.

- When I scroll in this tab the save and reboot button disappears.
Imo the button should stay at the bottom, as it is in the other tabs.

- The Motor direction dialog is also missing.

haslinghuis commented:
Thanks, the button is now fixed to the bottom like other tabs. The direction dialog button is not showing indeed, probably a CSS thingy I have to find out...

chmelevskij commented:
> @chmelevskij added my own recursive alternative one-liner (spread it out over multiple lines for readability) but `Promise.all` or `MULTIPLE_MSP` all have same outcome - we should rather focus on a BIOS like approach (Load once, save once, only have to deal with status polls)
> 
> ```
> const mspCommands = [
>     MSPCodes.MSP_SET_ARMING_CONFIG,
>     MSPCodes.MSP_SET_MOTOR_3D_CONFIG,
>     MSPCodes.MSP_SET_MOTOR_CONFIG,
>     MSPCodes.MSP_SET_ADVANCED_CONFIG,
>     MSPCodes.MSP_SET_FEATURE_CONFIG,
>     MSPCodes.MSP_SET_MIXER_CONFIG,
> ];
> 
> let commands = mspCommands.length - 1;
> 
> function saveSettings() {
>     MSP.send_message(
>         mspCommands[commands],
>         mspHelper.crunch(mspCommands[commands]),
>         false,
>         (commands--) ? saveSettings : onFinishedSave
>     );
> }
> 
> function onFinishedSave() {
>     MSP.send_message(MSPCodes.MSP_EEPROM_WRITE, false, false, onFinishedReboot);
> }
> ```
> 
> 👽

I'm a bit confused how is this different from sequential
```javascript
try {
  await MSP.promise(MSPCodes.MSP_SET_ARMING_CONFIG);
  await MSP.promise(MSPCodes.MSP_SET_MOTOR_3D_CONFIG);
  await MSP.promise(MSPCodes.MSP_SET_MOTOR_CONFIG);
  // the rest
  load_html();
} catch(error) {
  console.log(error)
}
```

The recursive function just executes in sequence. Looking at [promise](https://github.com/betaflight/betaflight-configurator/blob/0e3f3b460a9ec04ebf29485a50c8ae0db43c07b5/src/js/msp.js#L375) implementation, it resolves only when message is acknowledged, effectively doing the same thing.

I might be missing something though.

haslinghuis commented:
@TheIsotopes - improved the layout a lot - maybe also move `3D Mode` to `ESC/Motor Features` ?

See screenshots in the attached issue.

TheIsotopes commented:
@haslinghuis the servo tab is okay.

The other two tabs still need some cosmetic adjustments.

in the configuration tab:
The gaps between the boxes on the right are larger than on the left.
The boxes on the right should be align on both sides, front is ok but not at the end.

![config_tab](https://user-images.githubusercontent.com/29702171/103994007-b5161680-5196-11eb-9016-280cc7655cc3.jpg)


in the motors/esc tab:
I take a closer look, the comment will come later.

haslinghuis commented:
Thanks for the review! Back to work ;-)

haslinghuis commented:
@chmelevskij - I don't want to change too much here. We should address MSP (except polling) with a PG like solution. I had some side issues using the async specifier

haslinghuis commented:
@TheIsotopes I managed to fixed most of the issues. But it seems `grid-col col12` and 2x `grid-col col6` are not the same in different `grid-row` 's in motor tab (tried already to wrap in a `col12` but didn't work.

@Asizon ?

Asizon commented on src/css/tabs/motors.css:
```diff
@@ -1,3 +1,243 @@
+.tab-motors .gui_box {
+    margin-bottom: 10px;
+    float: left;
+    font-weight: bold;
```
Why added bold to these fonts?I think it should removed.

Asizon reviewed (changes requested):
Good job @haslinghuis now this motor tab looks really better.

Asizon commented:
@haslinghuis I made some adjustments to this box, what do you think?

![image](https://user-images.githubusercontent.com/43983086/104086285-d9392c80-5256-11eb-938f-729d4d7de772.png)

TheIsotopes commented:
> As this last issue is just cosmetic I think the work is finished. We can always fix this later.

@haslinghuis I agree, it's ok from my side.
But I would also include the last suggestion from @Asizon.

Asizon commented:
I forgot that we can also improve this part a bit, like here:

![image](https://user-images.githubusercontent.com/43983086/104086461-a09a5280-5258-11eb-999a-d5151204d1a8.png)

haslinghuis commented:
Thanks @Asizon for helping with styling the mixer, buttons and notice!!!

Asizon commented:
Good job, this part looks good on android
![Screenshot_20210110_171303_com betaflight betaflightconfigurator](https://user-images.githubusercontent.com/43983086/104128359-75eaef80-5367-11eb-8ef4-6d9ce29d372c.jpg)
![Screenshot_20210110_171255_com betaflight betaflightconfigurator](https://user-images.githubusercontent.com/43983086/104128361-76838600-5367-11eb-9e88-210c18802e19.jpg)

limonspb commented:
Very good job!!! I like the direction where it is moving.

Some small suggestions on visual style:

![image](https://user-images.githubusercontent.com/2925027/104134635-95c2e900-5350-11eb-94d5-be7c7a493140.png)
![image](https://user-images.githubusercontent.com/2925027/104134725-2b5e7880-5351-11eb-834d-ae84e00e122a.png)

haslinghuis commented:
Yes @limonspb like the last one. Now have 2 x col6. Will look at it in a hour.

haslinghuis commented:
Results:

![Screenshot from 2021-01-11 01-42-04](https://user-images.githubusercontent.com/8344830/104139925-e0734e00-53ae-11eb-87c8-ac67844641bb.png)

haslinghuis commented:
Tried some implementations of the `webkit` sliders, the examples are easy to follow but to implement with bars is another story (for me anyway). The bars looks nice but are duplicate in my opinion. Values already on the bottom so why need the visual bar. But this are just excuses I'm not able to solve the UX part of this. Changed width from 60 to 50% for motor bars and looks much better now, but elements where cropping over each other so changed it back as it was.

McGiverGim commented:
If you have something "that works" and is functional is enough. If it is not perfect, some other can take the work after this is merged and modify the user interface if needed to change the visual bar.

As I say always, is better go step by step with not too big PRs, if not we end with something very big, more difficult to review and with usually more bugs.

haslinghuis commented:
After testing everything found that I had to add MSP_ARMING_CONFIG in motors.js. All should be good now.

mikeller commented on src/css/tabs/servos.css:
```diff
@@ -188,3 +188,306 @@
+.tab-servos .servoblock {
+    margin-bottom: 0;
+    background-color: #ECECEC;
+    /*background-image: -webkit-linear-gradient(top, transparent, rgba(0, 0, 0, 0.15));*/
```
If we don't use this then it should be removed.

haslinghuis replied:
Removed

mikeller commented on src/js/tabs/configuration.js:
```diff
@@ -1,31 +1,15 @@
-'use strict';
```
Why was this removed?

haslinghuis replied:
Added (there is no reason I would have removed it)

mikeller commented on src/js/tabs/configuration.js:
```diff
@@ -664,7 +438,7 @@ TABS.configuration.initialize = function (callback, scrollPosition) {
             const MAX_DENOM = semver.gte(FC.CONFIG.apiVersion, "1.24.0") ? 16 : 8;
 
-            for (let denom = 1; denom <= MAX_DENOM; denom++) {
+            for (let denom = 1; denom < MAX_DENOM; denom++) {
```
Is this actually fixing a bug, or is this accidental?

haslinghuis replied:
There were problems with it during development (out of scope). Not anymore. Reverted.

mikeller commented on src/js/tabs/configuration.js:
```diff
@@ -1215,19 +782,23 @@ TABS.configuration.initialize = function (callback, scrollPosition) {
             self.analyticsChanges = {};
 
+            // fill some data
+            FC.GPS_CONFIG.auto_baud = $('input[name="gps_auto_baud"]').is(':checked') ? 1 : 0;
```
Where have the version checks for this gone?

haslinghuis replied:
Added

mikeller commented on src/js/tabs/motors.js:
```diff
@@ -472,42 +528,187 @@ TABS.motors.initialize = function (callback) {
+            'MULTISHOT',
+        ];
+
+        escProtocols.push('BRUSHED');
```
Where have the version checks for this gone?

haslinghuis replied:
Changed (using EscProtocols utils now)

mikeller commented on src/js/tabs/motors.js:
```diff
@@ -703,34 +936,89 @@ TABS.motors.initialize = function (callback) {
             }
         }
 
+        $('a.save').on('click', function() {
```
Since we are on the motors tab, what about making sure the motors are stopped before saving the motor configuration and rebooting the flight controller?

haslinghuis replied:
Added disarming before saving

mikeller replied:
Sorry, I think I am overlooking something - where do you stop the motors?

mikeller replied:
`setArmingEnabled()` will only disable arming and disarm (if armed). This does not affect motors spinning based on slider settings at all - see the handler for `#motorsEnableTestMode` for what needs to be done.

mikeller commented on src/js/tabs/servos.js:
```diff
@@ -41,49 +41,45 @@ TABS.servos.initialize = function (callback) {
-            servoHeader = servoHeader + '\
-                <th >A' + (i+1) + '</th>\
-            ';
+            servoHeader += `<th>A${(i+1)}</th>`;
```
Has the servo functionality been tested?

haslinghuis replied:
@TheIsotopes has tested this? Did you test it with an actual servo motor?

mikeller commented on src/js/tabs/motors.js:
```diff
@@ -703,34 +936,89 @@ TABS.motors.initialize = function (callback) {
+            let commands = mspCommands.length - 1;
+
+            function saveSettings() {
+                MSP.send_message(
```
I am sorry but this is very hard to understand - I think using `MSP.promise` will make this a lot easier to understand.
Is my guess right that the list in `mspCommands` is executed backwards? Why?
Also, n.b. that order matters for MSP commands setting configuration, as some of them have an overlap.

haslinghuis replied:
Changed to promise

mikeller reviewed (changes requested):
Not sure about this. Things that I have doubt about:
- not sure if it is a good approach to spread the configuration over more tabs than it already is spread over;
- the space usage in the motors tab is extremely bad - the 'sliders' are so comically small to be practically unusable, and the 'bars' are almost square - if we don't think that users should use the sliders we should remove them and replace them with number inputs. The mixer selection is now comically large compared to the sliders and bars;
- the panels in the configuration tab are now an unordered mess that lacks grouping - we should re-instate the grouping that we had before.

Things that are good:
- the servo output bars have been moved to the servos tab.

McGiverGim commented:
In my opinion is good to spread the configuration if it goes to the more intuitive place. And in my opinion, is more intuitive to group them by functionality, and motor is one.
I think the sliders can be better, but is difficult to make them thinner. They contain the telemetry data too. But I'm sure that once moved the servos, it will be easier to redo some things and make it look better. But I think is better to do it in another PR to not complicate this more.

haslinghuis commented on locales/en/messages.json:
```diff
@@ -1184,7 +1184,7 @@
     },
     "configurationFeaturesHelp": {
-        "message": "<strong>Note:</strong> Not all combinations of features are valid.  When the flight controller firmware detects invalid feature combinations conflicting features will be disabled.<br /><strong>Note:</strong> Configure serial ports <span class=\"message-negative\">before</span> enabling the features that will use the ports."
+        "message": "<strong>Note:</strong> Not all combinations of features are valid.  When the flight controller firmware detects invalid feature combinations conflicting features will be disabled.<br /><strong>Note:</strong> Configure serial ports <span class=\"message-negative\">before</span> enabling the features that will use the ports.<br /><strong>Note:</strong> Configuration of Mixer and ...
```
Removed

haslinghuis commented on src/css/tabs/motors.css:
```diff
@@ -11,7 +252,7 @@
 .tab-motors #graph {
     float: left;
-    width: calc(100% - 292px);
+    width: calc(100% - 160px); /* calc(100% - 292px); */
```
Removed

haslinghuis commented:
Actually  the bars on nightly look also square. I've added some media queries to change the width of the bars on a higher screen resolution. Mixer preview is 33% smaller now. Are the bars really necessary, sliders already show the motor value below them.

haslinghuis commented:
Okay used a Makeblock arduino servo to test the servo test and it works !!! (I only lack VBAT power on my F7 Testcontroller so the servo didn't move) - but the bars are working nice on RC input.

mikeller commented on src/js/tabs/motors.js:
```diff
@@ -681,56 +896,93 @@ TABS.motors.initialize = function (callback) {
+
+            const saveSettings = async () => {
+                try {
+                    await MSP.promise(MSPCodes.MSP_SET_FEATURE_CONFIG, mspHelper.crunch(MSPCodes.MSP_SET_FEATURE_CONFIG));
```
Using a block of consecutive awaits on individual promises is completely against the nature of promises, and will likely lead to problems and a frozen UI due to the main thread being blocked for a long time.
Instead, use promise chaining in the way that MSP promises are used in other places in the application, like `src/js/tabs/pid_tuning.js` to get the MSP communication done in a background thread.

haslinghuis replied:
$('a.save').on('click', function() {
            mspHelper.setArmingEnabled(false, true);

This is stopping the motors.
Have tested save and reboot many times but will adapt the method used in pid_tuning...

haslinghuis replied:
Done.

mikeller commented:
@haslinghuis: Thanks for testing the servo functionality.

@McGiverGim: I am not sure that spreading out the configuration settings into more and more tabs (that prior to this did not exist or did not contain any configuration settings is going to make this application easier to use, or just more confusing. This is especially the case when settings that have interdependencies (like gyro loop speeds and motor protocol settings) are moved to different tabs - users now have to know that they have to flip between these tabs to get this configured properly.
But if it is the consensus that this is making the application better then we can merge this.
But I think we will have to make an effort to clean up the configuration tab, and give it back some semblance of a logical ordering before we consider the next release.

haslinghuis commented:
@mikeller work is already underway after this to move receiver configuration to receiver tab.

Motor tab is setting correct pid setting on configuration tab depending on dshot protocol chosen already now.

There are no feature requests for the following so I just throw in some wild ideas:
Was thinking about moving gyro and sensors to sensor tab and gps configuration to gps. Craft name and camera angle should be in osd. Dshot beepers maybe to motors also and the features that are left we have to see what's is left and beeper configuration also to receiver. The features that are left can go to their respective tab.

mikeller commented:
@haslinghuis:

> Motor tab is setting correct pid setting on configuration tab depending on dshot protocol chosen already now.

Not sure what you want to say here - the fact is that currently users do not need to use the motor tab to set up the basic configuration for their craft, and your pull request changes this, making it harder to get this task done.

> There are no feature requests for the following so I just throw in some wild ideas:

As I said, in my opinion we have to be careful with this - I think there should be a limited number of tabs that a user is required to use in order to get the basic configuration of their craft done. Adding the requirement to use more tabs (like what we are doing in this pull request) will just make the application harder to use for this very basic task.
Otherwise, especially for new users, getting a craft to a point where it is usable will become very tedious, as they will have to explore through a large number of tabs that they are not familiar with in order to find all the places where they are required to enter information.

haslinghuis commented:
@mikeller did you look at the bars? Are they better now?

mikeller commented:
@haslinghuis: Yes, this looks a bit better. I still think that, considering how much space we have available on this tab, it should be possible to come up with a design that allows us to have the sliders in a length that actually makes them usable (even on a mobile device).

But the matter of properly stopping the motors before changing settings on the flight controller and then saving is still open.

haslinghuis commented:
Now it is stopping motors on save only. Indeed we should disable testing when changing settings. Keep state so until saved we cannot test (could show a dialog to save first). This is also true for the configuration of reordering and direction buttons. Also while testing we should not be allowed to change settings. Just thinking how to best implement this like I'm changing save / save and reboot button in the receiver PR I 'm working on.

haslinghuis commented:
Have been very busy today. Have some experimental code for the button which could be removed in no time, but I just want to show it. Motors are disabled in proper manner now in all situations.

haslinghuis commented:
Fixed spacing

haslinghuis commented:
Reverted to initial design (thanks to @Asizon to ditch the mixer preview) - everything in one page now.

haslinghuis commented:
![Screenshot from 2021-01-20 07-15-58](https://user-images.githubusercontent.com/8344830/105134964-7fa2ee80-5aef-11eb-8180-b2e960a9874b.png)

McGiverGim commented:
I now here we have a lot of opinions about how it must look... but here one more, my opinion :)

First, I think your latest capture is good enough, so this is only to let some thinking...

- The configurations motor fields can grow with new versions, so it must be something that can grow in space without problem.
- The gyro information maybe it does not need to be very long. Maybe with the size of the test bars is enough.
- I liked when the craft was in the configuration of the mixer. I think it does nothing with the gyro. It moves or shows movement of the motors? I don't remember...

So in my opinion:
- The configuration must go first, to let it grow easy. And in reduced size screens (like Android) it will be shown when entering in the tab. We are changing it from place and maybe someone does not find it if it goes under the test.
- Move the craft to the mixer from the gyro.
- Move the gyro over the test sliders with only the size of the test sliders.

But like I said, there are many opinions, so I will not change nothing, your latest capture is good for a first version and we can move/change depending on some feedback. If we continue changing it with each opinion we will not end never...

haslinghuis commented:
Thanks @McGiverGim - this really looks good (didn't push yet):

![Screenshot from 2021-01-21 01-52-02](https://user-images.githubusercontent.com/8344830/105259119-78c9b980-5b8b-11eb-81cd-fa6d1f511b64.png)

haslinghuis commented:
Fixed the mixer tab too.

![Screenshot from 2021-01-21 05-11-53](https://user-images.githubusercontent.com/8344830/105279146-3d3ce880-5ba7-11eb-9932-2da0bd278cba.png)

haslinghuis commented:
Thanks @McGiverGim - your review was very helpfull. Addressed your points and it should be good to go now:

![Screenshot from 2021-01-21 05-11-53](https://user-images.githubusercontent.com/8344830/105280666-922e2e00-5baa-11eb-88f5-94902604aa7a.png)

McGiverGim reviewed (dismissed):
LGTM, but as I said, others will not think the same :)

haslinghuis commented:
Minor improvement on width in motor testing (E values were too condensed) and moved @limonspb buttons to the content_toolbar. The mixer section does also comply with the ESC/Motor Features.

@mikeller having addressed stopping of motors properly when changing settings please take a look again.

haslinghuis commented:
@Asizon did you find a way to move the sliders beside the bars? Have tried but didn't work.

haslinghuis commented:
Maybe should hide the motor direction and reorder motors buttons also when testing is active.

haslinghuis commented:
Apart from the motor testing part (bars / sliders) - which should be done separate this should be really be done now. Until requested I am not going to change anything more.

McGiverGim commented:
@mikeller can we merge this? It seems finished to me. Some change is pending?

mikeller commented:
@McGiverGim: I'll review. This has had so many repetitive changes pushed into it that I stopped looking at it.

haslinghuis commented:
👍

limonspb commented:
Congratulations, @haslinghuis very good job! :)
